TypeScript_ESLint.snb.md

```diff
@@ -2,92 +2,176 @@
-    ForInStatement: (node): void => { ... }
-}
-```
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-extra-non-null-assertion.ts
+
+Note that rules declared in the TypeScript ESLint project use a `util.createRule` helper that internally calls to an `ESLintUtils.RuleCreator`:
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/util/createRule.ts
+
+`RuleCreator` adds a documentation URL to rules under `meta.docs.url` based on their name.
+TypeScript ESLint rules are available under https://typescript-eslint.io/rules, including https://typescript-eslint.io/rules/no-extra-non-null-assertion.
+
+### Rule Metadata Properties
+
+Two particularly useful rule metadata properties are specified in many TypeScript ESLint:
+
+- `fixable`: If the rule includes an auto-fixer on any of its messages, either `"code"` or `"whitespace"`
+- `messages`: An object keying ID strings to full messages that may be reported by the rule
+
+`no-extra-non-null-assertion`'s `meta` indicates it has an auto-fixer for code issues, and that it may report _`Forbidden extra non-null assertion`_ if it detects an issue.
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-extra-non-null-assertion.ts?L6-17
+
+### Rule Creator Functions
+
+A rule's `create` function returns an object whose keys are AST queries and whose values are functions that will be run whenever a node matching the corresponding query is found.
+The AST query keys use the [ESQuery](https://github.com/estools/esquery) library, which allows using syntax similar to CSS selectors for querying types of nodes.
+
+`no-extra-non-null-assertion` creates three node queries in its returned object:
+
+1. Any `!` non-null assertion within another `!` non-null assertion: e.g. `scream!!.length`
+2. Any `!` non-null assertion just before an `?.` optional chained property: e.g. `scream!?.length`
+3. Any `!` non-null assertion just before an `?.` optional chained call: e.g. `scream!?.()`
 
-This key will be used to match this rule with the name of an AST node type in the ESLint AST, so that the node argument will always have the type that the rule targets:
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-extra-non-null-assertion.ts?L19-39
 
-https://sourcegraph.com/npm/typescript-eslint/types@540cca215af38ab1a49b6cde8bcb492f4b6e53c5/-/blob/dist/generated/ast-spec.d.ts?L75
+Note that any AST node names prefixed with `TS`, including `TSNonNullExpression`, are TypeScript-specific nodes.
+These are available to ESLint rules when using [`@typescript-eslint/parser`](http://npmjs.com/package/@typescript-eslint/parser) instead of ESLint's built-in parser.
+The `@typescript-eslint/utils` package exports those types under the `TSESTree` namespace.
+
+⚠ TODO: link to docs page for all of them? ⚠
```

Comment:
@bradzacher you've been working on https://github.com/typescript-eslint/typescript-eslint/issues/2726 recently. Do you think we'll have something we can include on the website in time for this blog post?

If not, we can go back to linking to `ast-spec.d.ts` like it was before.

Reply:
when is this blog post slated to release?
I'd say *probably not* as I'm kind of bandwidth-constrained ATM, but depends on the lead time!

Reply:
I went back to the `ast-spect.d.ts` reference instead. Let's just get this out the door 😄